sympy/combinatorics/perm_groups.py

```diff
@@ -1702,6 +1705,78 @@ def is_abelian(self):
                     return False
         return True
 
+    def abelian_invariants(self):
+        """
+        Returns the abelian invariants for the given group.
+        Let ``G`` be a nontrivial finite abelian group. Then G is isomorphic to
+        the direct product of finitely many nontrivial cyclic groups of
+        prime-power order.
+
+        The prime-powers that occur as the orders of the factors are uniquely
+        determined by G. More precisely, the primes that occur in the orders of the
+        factors in any such decomposition of ``G`` are exactly the primes that divide
+        ``|G|`` and for any such prime ``p``, if the orders of the factors that are
+        p-groups in one such decomposition of ``G`` are ``p^{t_1} >= p^{t_2} >= ... p^{t_r}``,
+        then the orders of the factors that are p-groups in any such decomposition of ``G``
+        are ``p^{t_1} >= p^{t_2} >= ... p^{t_r}``.
+
+        The uniquely determined integers ``p^{t_1} >= p^{t_2} >= ... p^{t_r}``, taken
+        for all primes that divide ``|G|`` are called the invariants of the nontrivial
+        group ``G`` as suggested in ([14], p. 542).
+
+        Notes
+        =====
+
+        We adopt the convention that the invariants of a trivial group are [].
+
+        Examples
+        ========
+
+        >>> from sympy.combinatorics import Permutation
+        >>> from sympy.combinatorics.perm_groups import PermutationGroup
+        >>> a = Permutation([0, 2, 1])
+        >>> b = Permutation([1, 0, 2])
+        >>> G = PermutationGroup([a, b])
+        >>> G.abelian_invariants()
+        [2]
+        >>> from sympy.combinatorics.named_groups import CyclicGroup
+        >>> G = CyclicGroup(7)
+        >>> G.abelian_invariants()
+        [7]
+
+        """
+        if self.is_trivial:
+            return []
+        gns = self.generators
+        inv = []
+        der = self.derived_subgroup()
+        G = self
+
+        for p in primefactors(G.order()):
+            ranks = []
+            r = 2
+            while r > 1:
+                H = der
+                for g in gns:
+                    elm = g**p
+                    if not H.contains(elm):
```

Comment:
`contains` is a fairly expensive method since a coset factorization has to be computed. I think that the test should be taken with the same (fixed) `H` for all powers `g**p`. That filtered list of powers could probably also used as `gens` for the next round of the `r` loop instead of `G.generators` that is actually `H.generators` and thus contains also generators of `der`.

Reply:
By fixed `H` you mean we should avoid `H = der`? and use the same `H` as computed in the previous iteration.

Reply:
I mean that `H` should not be redefined for each `g in gns` (whose power is not contained in it). Instead, we should collect those powers `g**p` that are not in `H` in a list and then update `H` by adding all those new generators at once instead of one at a time. (So `H` is fixed during one for loop.)

Reply:
Okay, I got it!

Reply:
Done!, please have a look.